File: src/view/renderer/RendererImplConsole.java
package view.renderer;

import model.structs.Point;
import model.structs.Rect;
import view.console.*;
import view.components.AbstractDisplayBlock;
import view.components.DisplayBlock;
import view.components.GridBlock;
import view.components.TextBlock;

public class RendererImplConsole implements Renderer {
    private final Screen screen;
    private final Window window;
    public static final Rect WINDOW_RECT = new Rect(0,32,0,120);

    protected RendererImplConsole() {
        screen = new ScreenImplConsole(WINDOW_RECT.bottom,WINDOW_RECT.right);
        window = screen.createWindow(WINDOW_RECT);
    }

    private boolean recursiveRender(AbstractDisplayBlock view, Rect availableRect) {
        Rect displayRect = new Rect(
                availableRect.top + view.getMargin().top,
                availableRect.bottom - view.getMargin().bottom,
                availableRect.left + view.getMargin().left,
                availableRect.right - view.getMargin().right);

        if(displayRect.top >= displayRect.bottom
                || displayRect.left >= displayRect.right) {
            return false;
        }

        if(view.getAvailableWidth() > displayRect.right - displayRect.left) { return false; }
        switch (view.getHorizontalAlign()) {
            case BEGIN -> displayRect.right = displayRect.left + view.getAvailableWidth();
            case END -> displayRect.left = displayRect.right - view.getAvailableWidth();
            case MIDDLE -> {
                displayRect.left = (displayRect.right + displayRect.left - view.getAvailableWidth())/2;
                displayRect.right = displayRect.left + view.getAvailableWidth();
            }
        }

        if(view.getAvailableHeight() > displayRect.bottom - displayRect.top) { return false; }
        switch (view.getVerticalAlign()) {
            case BEGIN -> displayRect.bottom = displayRect.top + view.getAvailableHeight();
            case END -> displayRect.top = displayRect.bottom - view.getAvailableHeight();
            case MIDDLE -> {
                displayRect.top = (displayRect.top + displayRect.bottom - view.getAvailableHeight())/2;
                displayRect.bottom = displayRect.top + view.getAvailableHeight();
            }
        }

        Point startPoint = new Point(displayRect.left, displayRect.top);
        if (view instanceof TextBlock viewText) {
            window.print(startPoint, viewText.getText());
            return true;
        } else if (view instanceof GridBlock viewGrid) {
            int width = viewGrid.getGrid()[0].length;
            int height = viewGrid.getGrid().length;
            Point offset = new Point();
            Point showPosition = new Point(startPoint.x, startPoint.y);
            for(offset.x = 0; offset.x < width; offset.x++) {
                showPosition.y = startPoint.y;
                for(offset.y = 0; offset.y < height; offset.y++) {
                    window.setPixel(showPosition, new PixelImplConsole((char)(viewGrid.getGrid()[offset.y][offset.x])));
                    showPosition.y++;
                }
                showPosition.x += 2;
            }
            return true;
        } else if (view instanceof DisplayBlock viewDiv) {
            switch(viewDiv.getChildLayout()) {
                case FLEX_X -> {
                    int innerMargin = viewDiv.getAvailableWidth();
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        innerMargin -= child.getOccupiedWidth();
                    }
                    innerMargin /= viewDiv.getChildren().size() - 1;
                    Rect childrenRect = new Rect(displayRect.top,displayRect.bottom,displayRect.left,displayRect.right);
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        childrenRect.right = childrenRect.left + child.getOccupiedWidth();
                        if(!recursiveRender(child, childrenRect))
                            return false;
                        childrenRect.left = childrenRect.right + innerMargin;
                    }
                }
                case FLEX_Y -> {
                    int innerMargin = viewDiv.getAvailableHeight();
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        innerMargin -= child.getOccupiedHeight();
                    }
                    innerMargin /= viewDiv.getChildren().size() - 1;
                    Rect childrenRect = new Rect(displayRect.top,displayRect.bottom,displayRect.left,displayRect.right);
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        childrenRect.bottom = childrenRect.top + child.getOccupiedWidth();
                        if(!recursiveRender(child, childrenRect))
                            return false;
                        childrenRect.top = childrenRect.bottom + innerMargin;
                    }
                }
                case LEFT_TO_RIGHT -> {
                    Rect childrenRect = new Rect(displayRect.top,displayRect.bottom,displayRect.left,displayRect.right);
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        childrenRect.right = childrenRect.left + child.getOccupiedWidth();
                        if(!recursiveRender(child, childrenRect))
                            return false;
                        childrenRect.left = childrenRect.right;
                    }
                }
                case RIGHT_TO_LEFT -> {
                    Rect childrenRect = new Rect(displayRect.top,displayRect.bottom,displayRect.left,displayRect.right);
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        childrenRect.left = childrenRect.right - child.getOccupiedWidth();
                        if(!recursiveRender(child, childrenRect))
                            return false;
                        childrenRect.right = childrenRect.left;
                    }
                }
                case UP_TO_DOWN -> {
                    Rect childrenRect = new Rect(displayRect.top,displayRect.bottom,displayRect.left,displayRect.right);
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        childrenRect.bottom = childrenRect.top + child.getOccupiedHeight();
                        if(!recursiveRender(child, childrenRect))
                            return false;
                        childrenRect.top = childrenRect.bottom;
                    }
                }
                case DOWN_TO_UP -> {
                    Rect childrenRect = new Rect(displayRect.top,displayRect.bottom,displayRect.left,displayRect.right);
                    for(AbstractDisplayBlock child : viewDiv.getChildren()) {
                        childrenRect.top = childrenRect.bottom - child.getOccupiedHeight();
                        if(!recursiveRender(child, childrenRect))
                            return false;
                        childrenRect.bottom = childrenRect.top;
                    }
                }
            }
        }
        return true;
    }

    /**
     * @param view the view
     * @return true if succeeded
     */
    @Override
    public boolean render(AbstractDisplayBlock view) {
        window.clearWindowBuffer();
        if(!recursiveRender(view, screen.getRect()))
            return false;
        window.paint();
        screen.paint();
        return true;
    }
}

File: src/view/renderer/RendererImplConsole.kt
package view.renderer

import model.structs.Point
import model.structs.Rect
import view.components.*
import view.components.AlignType.*
import view.console.PixelImplConsole
import view.console.Screen
import view.console.ScreenImplConsole
import view.console.Window

class RendererImplConsole : Renderer {
    private val screen: Screen
    private val window: Window

    init {
        screen = ScreenImplConsole(WINDOW_RECT.bottom, WINDOW_RECT.right)
        window = screen.createWindow(WINDOW_RECT)
    }

    private fun recursiveRender(block: AbstractDisplayBlock, availableRect: Rect): Boolean {
        val displayRect = Rect(
            availableRect.top + block.margin.top,
            availableRect.bottom - block.margin.bottom,
            availableRect.left + block.margin.left,
            availableRect.right - block.margin.right
        )

        if (displayRect.top >= displayRect.bottom
            || displayRect.left >= displayRect.right
        ) {
            return false
        }

        if (block.availableWidth > displayRect.right - displayRect.left) {
            return false
        }
        when (block.horizontalAlign) {
            BEGIN -> displayRect.right = displayRect.left + block.availableWidth
            END -> displayRect.left = displayRect.right - block.availableWidth
            MIDDLE -> {
                displayRect.left = (displayRect.right + displayRect.left - block.availableWidth) / 2
                displayRect.right = displayRect.left + block.availableWidth
            }

            null -> throw IllegalArgumentException("NULL should not happen")
        }

        if (block.availableHeight > displayRect.bottom - displayRect.top) {
            return false
        }
        when (block.verticalAlign) {
            BEGIN -> displayRect.bottom = displayRect.top + block.availableHeight
            END -> displayRect.top = displayRect.bottom - block.availableHeight
            MIDDLE -> {
                displayRect.top = (displayRect.top + displayRect.bottom - block.availableHeight) / 2
                displayRect.bottom = displayRect.top + block.availableHeight
            }

            null -> throw IllegalArgumentException("NULL should not happen")
        }

        val startPoint = Point(displayRect.left, displayRect.top)
        if (block is TextBlock) {
            window.print(startPoint, block.getText())
            return true
        } else if (block is GridBlock) {
            val width = block.grid[0].size
            val height = block.grid.size
            val offset = Point()
            val showPosition = Point(startPoint.x, startPoint.y)
            offset.x = 0
            while (offset.x < width) {
                showPosition.y = startPoint.y
                offset.y = 0
                while (offset.y < height) {
                    window.setPixel(showPosition, PixelImplConsole((block.grid[offset.y][offset.x]).toChar()))
                    showPosition.y++
                    offset.y++
                }
                showPosition.x += 2
                offset.x++
            }
            return true
        } else if (block is DisplayBlock) {
            when (block.childLayout) {
                ChildLayout.FLEX_X -> {
                    var innerMargin = block.availableWidth
                    for (child in block.children) {
                        innerMargin -= child.occupiedWidth
                    }
                    innerMargin /= block.children.size - 1
                    val childrenRect = Rect(displayRect.top, displayRect.bottom, displayRect.left, displayRect.right)
                    for (child in block.children) {
                        childrenRect.right = childrenRect.left + child.occupiedWidth
                        if (!recursiveRender(child, childrenRect)) return false
                        childrenRect.left = childrenRect.right + innerMargin
                    }
                }

                ChildLayout.FLEX_Y -> {
                    var innerMargin = block.availableHeight
                    for (child in block.children) {
                        innerMargin -= child.occupiedHeight
                    }
                    innerMargin /= block.children.size - 1
                    val childrenRect = Rect(displayRect.top, displayRect.bottom, displayRect.left, displayRect.right)
                    for (child in block.children) {
                        childrenRect.bottom = childrenRect.top + child.occupiedWidth
                        if (!recursiveRender(child, childrenRect)) return false
                        childrenRect.top = childrenRect.bottom + innerMargin
                    }
                }

                ChildLayout.LEFT_TO_RIGHT -> {
                    val childrenRect = Rect(displayRect.top, displayRect.bottom, displayRect.left, displayRect.right)
                    for (child in block.children) {
                        childrenRect.right = childrenRect.left + child.occupiedWidth
                        if (!recursiveRender(child, childrenRect)) return false
                        childrenRect.left = childrenRect.right
                    }
                }

                ChildLayout.RIGHT_TO_LEFT -> {
                    val childrenRect = Rect(displayRect.top, displayRect.bottom, displayRect.left, displayRect.right)
                    for (child in block.children) {
                        childrenRect.left = childrenRect.right - child.occupiedWidth
                        if (!recursiveRender(child, childrenRect)) return false
                        childrenRect.right = childrenRect.left
                    }
                }

                ChildLayout.UP_TO_DOWN -> {
                    val childrenRect = Rect(displayRect.top, displayRect.bottom, displayRect.left, displayRect.right)
                    for (child in block.children) {
                        childrenRect.bottom = childrenRect.top + child.occupiedHeight
                        if (!recursiveRender(child, childrenRect)) return false
                        childrenRect.top = childrenRect.bottom
                    }
                }

                ChildLayout.DOWN_TO_UP -> {
                    val childrenRect = Rect(displayRect.top, displayRect.bottom, displayRect.left, displayRect.right)
                    for (child in block.children) {
                        childrenRect.top = childrenRect.bottom - child.occupiedHeight
                        if (!recursiveRender(child, childrenRect)) return false
                        childrenRect.bottom = childrenRect.top
                    }
                }
            }
        }
        return true
    }

    /**
     * @param block the block
     * @return true if succeeded
     */
    override fun render(block: AbstractDisplayBlock): Boolean {
        window.clearWindowBuffer()
        if (!recursiveRender(block, screen.rect)) return false
        window.paint()
        screen.paint()
        return true
    }

    companion object {
        val WINDOW_RECT: Rect = Rect(0, 32, 0, 120)
    }
}
